### ISSUE TITLE
Hed

### ISSUE THREAD

**Issue by VoyageWang** [khdlr/HED-UNet#12]
I am confused about how you do the attention with the softmax? Is it in the channel dimension?
why it can be treated as attention? How can we use the code to do segmentation?

**Comment by khdlr** [comment]
Kindly refer to Fig. 3 in the paper (https://arxiv.org/pdf/2103.01849.pdf). The attention is not in the channel dimension, but in the "resolution" dimension. There is only one query (the output), and one key/value for each resolution level.

In order to use it for your segmentation project, you're going to need to adapt `data_loading.py` for your own dataset and then simply call `python train.py`.

**Comment by VoyageWang** [comment]
你好，你的邮件已经被我接受。hello，your email has been recieved. Thank you.

**Comment by VoyageWang** [comment]
Thanks for your reply! I had met another question about how to visualize the effective receptive field like fig9 in the article. Can you show me the code? This can help me a lot. Thanks again!

**Comment by khdlr** [comment]
The general idea is to backpropagate the output of a central pixel, and average that over your validation dataset:

```python
erfs = []

for img, mask in val_loader:
  img = img.to(device)
  img.requires_grad = True
  out = model(img)
  _, _, H, W = out.shape
  center = out[:, :, H//2, W//2].mean()
  center.backward()
  erfs.append(img.grad.abs().mean(dim=[1,2]))

erf = torch.mean(torch.stack(erfs, dim=0), dim=0)
```

Written from memory - I am not on the server to double-check the accuracy of the above code. Let me know if you have any additional questions.

**Comment by VoyageWang** [comment]
Thanks for your reply! I got this idea and understood the idea of ERF after your teaching. I have another question about how to get the figure of edge like fig 5. In other words, how to detect the edge from the segmentation results? Can you show me the code?

**Comment by khdlr** [comment]
The model directly generates an edge detection map, there is no need to derive the edges from the segmentation results. I.e. `model(inputs)[batch_idx, 0]` will be the segmentation prediction and `model(inputs)[batch_idx, 1]` is the edge prediction.

The function used for the plots in fig. 5 is literally the first function in `train.py`: https://github.com/khdlr/HED-UNet/blob/49f8d3005e79ed90765e46daa0a76cfb6bebe424/train.py#L41

**Comment by VoyageWang** [comment]
你好，你的邮件已经被我接受。hello，your email has been recieved. Thank you.
